### ISSUE TITLE
Apply throttling to writes when hints are waiting to be written to disk

### ISSUE THREAD

**Issue by piodul** [scylladb/scylla#8139]
Hinted handoff uses a simple overload protection mechanism on the write path, designed to limit the amount of RAM used by hints. At the beginning of a write operation, coordinator looks at the total size of hints currently waiting to be admitted to the hints' commitlog. If their size is greater than 10MB, the coordinator immediately dismisses the write request with the OverloadedException response.

This mechanism is similar to the one Cassandra has; C* limits the _number_ of hints waiting to be admitted to HintsStore (their counterpart of our hints' commitlog). The limit is currently equal to 128 * available processors.

The current scheme causes some trouble in our rolling upgrade tests. The pattern is very simple: after the first node is shut down for upgrade, other nodes start generating hints on every write. Nodes do not keep up with flushing hints to disk, so the 10MB buffer fills up which causes regularly occuring periods of time when nodes reject writes. This breaks some of our test as cassandra-stress gives up after receiving the OverloadedException 10 times in a row. This situation can be observed in some of the hinted handoff issues:

#6438 #6964 #7112

We should improve the current situation with overload protection for hints. For example, we could apply throttling before we start rejecting writes due to hints.

@slivne proposed the following idea: on the coordinator, write handlers should wait for hints to be written before being deleted. Hints wouldn't count towards consistency (unless CL=ANY is used), but they would cause the writes to wait in the background if hints are a bottleneck. This will allow us to use the throttling mechanism described in [this blog post](https://www.scylladb.com/2018/12/04/worry-free-ingestion-flow-control/). I like this idea because it plugs into an existing overload protection framework. Furthermore, we wouldn't need to additionally account for the space hints take up in RAM, because it is already accounted for by the write operation - now, hint waiting to be written can live longer than the write operation it originated from, so it needs to prolong the life of the mutation and it makes sense to additionally worry about the memory taken up by hints.

cc: @vladzcloudius @gleb-cloudius @avikivity @slivne @psarna @haaawk what do you think about this idea? Do you have any other ideas?

**Comment by avikivity** [comment]
Good idea. I'm surprised it wasn't already like that, launching an operation in the background is a recipe for disaster (and the write throttling mechanism we have attempts to mitigate the disaster we have from the post-CL writes).

**Comment by vladzcloudius** [comment]
I agree with @avikivity. Good idea.

**Comment by psarna** [comment]
Ref: https://github.com/scylladb/scylla/issues/6069
